be/src/runtime/runtime_filter_cache.cpp

```diff
@@ -0,0 +1,217 @@
+    RfCacheValue() { _extend_lifetime(); }
+    ~RfCacheValue() = default;
+    void put_if_absent(int filter_id, const JoinRuntimeFilterPtr& filter) {
+        _extend_lifetime();
+        if (!_filters.count(filter_id)) {
+            _filters[filter_id] = filter;
+        }
+    }
+
+    JoinRuntimeFilterPtr get(int filter_id) const {
+        _extend_lifetime();
+        auto it = _filters.find(filter_id);
+        if (it != _filters.end()) {
+            return it->second;
+        } else {
+            return nullptr;
+        }
+    }
+
+    bool is_expired() const {
+        auto now = duration_cast<milliseconds>(steady_clock::now().time_since_epoch()).count();
+        return now > _deadline;
+    }
+
+private:
+    void _extend_lifetime() const {
+        _deadline = duration_cast<milliseconds>(steady_clock::now().time_since_epoch() + EXPIRE_SECONDS).count();
+    }
+    std::unordered_map<int, std::shared_ptr<vectorized::JoinRuntimeFilter>> _filters;
+    mutable int64_t _deadline;
+    static constexpr seconds EXPIRE_SECONDS = seconds(60);
+};
+
+class RfEvent {
+public:
+    RfEvent(const TUniqueId& query_id, int filter_id, std::string&& msg)
+            : query_id(query_id),
+              filter_id(filter_id),
+              ts(duration_cast<milliseconds>(steady_clock::now().time_since_epoch()).count()),
+              msg(std::move(msg)) {}
+    ~RfEvent() = default;
+    std::string to_string() { return strings::Substitute("[$0][$1] $2: $3", print_id(query_id), filter_id, ts, msg); }
+
+private:
+    TUniqueId query_id;
+    int filter_id;
+    long ts;
+    std::string msg;
+};
+using RfEventPtr = std::shared_ptr<RfEvent>;
+using RfEvents = std::list<RfEventPtr>;
+class RfEventValue {
+public:
+    RfEventValue() = default;
+    ~RfEventValue() = default;
+    void add_rf_event(const TUniqueId& query_id, int filter_id, std::string&& msg) {
+        _extend_lifetime();
+        _events.push_back(std::make_shared<RfEvent>(query_id, filter_id, std::move(msg)));
+    }
+
+    std::list<std::string> get_events() {
+        _extend_lifetime();
+        std::list<std::string> event_strings;
+        for (auto it = _events.begin(); it != _events.end(); ++it) {
+            event_strings.push_back((*it)->to_string());
+        }
+        return event_strings;
+    }
+    bool is_expired() const {
+        auto now = duration_cast<milliseconds>(steady_clock::now().time_since_epoch()).count();
+        return now > _deadline;
+    }
+
+private:
+    void _extend_lifetime() const {
+        _deadline = duration_cast<milliseconds>(steady_clock::now().time_since_epoch() + EXPIRE_SECONDS).count();
+    }
+    RfEvents _events;
+    mutable int64_t _deadline;
+    static constexpr seconds EXPIRE_SECONDS = seconds(600);
+};
+
+RuntimeFilterCache::RuntimeFilterCache(size_t log2_num_slots)
+        : _num_slots(1L << log2_num_slots),
+          _slot_mask(_num_slots - 1),
+          _mutexes(_num_slots),
+          _filter_maps(_num_slots),
```

Comment:
Is this data structure complicated? Why cut to many slots? Have you benchmarked this code and see performance degradation because of this code?

Reply:
use slotted tables to reduce lock contention is a common method.